// @devvit/linkers/test/test-builder.js
import path from 'node:path';
import { NonNull } from '@devvit/shared-types/NonNull.js';
import { createRequire } from 'module';
const require = createRequire(import.meta.url);
export async function build(builder, actorDir) {
    const compileResponse = await compile(builder, actorDir);
    return await link(builder, compileResponse);
}
async function compile(builder, actorDir) {
    const filename = path.join(path.dirname(require.resolve('@devvit/internal-examples')), actorDir);
    const request = {
        config: undefined,
        info: {
            name: filename.split(path.sep).at(-1) ?? '',
            owner: 'snoo',
            version: '0.0.0',
        },
        minify: 'None',
        includeMetafile: false,
        root: filename,
    };
    const response = await builder.pack.compile(request);
    expect(response.errors).toHaveLength(0);
    expect(response.warnings).toHaveLength(0);
    expect(response.bundles).toHaveLength(2);
    return response;
}
async function link(builder, compileResponse) {
    const linkResponse = await builder.linker.link(compileResponse.bundles[0]);
    expect(linkResponse.errors).toHaveLength(0);
    return NonNull(linkResponse.bundle);
}
