// @devvit/client/effects/share.js
import { EffectType } from '@devvit/protos/json/devvit/ui/effects/v1alpha/effect.js';
import { emitEffect } from '@devvit/shared-types/client/emit-effect.js';
import { maxShareParamUserDataChars } from '@devvit/shared-types/client/share.js';
import { ICON_FILE_PATH } from '@devvit/shared-types/constants.js';
// to-do: move to web view scripts.
/** Show the native share sheet or copy to clipboard depending on device. */
export async function showShareSheet(opts) {
    if (opts.data && opts.data.length > maxShareParamUserDataChars)
        throw Error(`data must be <= ${maxShareParamUserDataChars} characters but was ${opts.data.length} characters`);
    const iconURL = `${new URL(ICON_FILE_PATH, location.origin)}`;
    let iconRsp;
    try {
        iconRsp = await fetch(iconURL, { method: 'HEAD' });
    }
    catch {
        //
    }
    // Create the share URL using the post from the props, or the current post if not provided.
    const postId = opts.post ?? devvit.context.postId;
    const postWithoutPrefix = postId.replace('t3_', '');
    const url = `https://reddit.com/r/_/comments/${postWithoutPrefix}`;
    emitEffect({
        type: EffectType.EFFECT_WEB_VIEW,
        share: {
            appIconUri: iconRsp?.ok ? iconURL : undefined,
            userData: opts.data,
            text: opts.text,
            title: opts.title,
            url: url,
        },
    });
}
export function getShareData() {
    return devvit.share?.userData;
}
