// @devvit/web-view-scripts/devvit-global.js
import { Client as ClientName, } from '@devvit/protos/json/devvit/ui/effects/web_view/v1alpha/context.js';
import { T2, T3, T5 } from '@devvit/shared-types/tid.js';
import {} from '@devvit/shared-types/webbit.js';
import { queryClientVersion } from './client.js';
import { decodeToken, requestTokenRefresh } from './token.js';
/** @internal */
export const initDevvitGlobal = (document, location, window) => {
    const bridge = getBridgeContext(window.name, location.hash);
    const client = getClient(bridge);
    let appPermissionState = bridge?.appPermissionState;
    /**
     * Android versions older than this do not support runAsUser correctly, so we
     * suppress the permission state to allow the app to run as if the feature
     * wasn't present (failing open).
     *
     * Supported after 2026.05.x
     */
    if (client?.name === 'ANDROID') {
        if (client.version.yyyy < 2026 ||
            (client.version.yyyy === 2026 && client.version.release <= 5)) {
            appPermissionState = undefined;
        }
    }
    const reqCtx = decodeToken(bridge.signedRequestContext);
    let context;
    if (reqCtx)
        context = contextFromRequestContext(reqCtx, client, bridge.postData);
    else if (bridge.webViewContext)
        context = contextFromWebViewContext(bridge.webViewContext, client, bridge.postData);
    if (!context)
        throw Error('no context');
    // to-do: expose `devvitDebug`, `shredditVersion`.
    const clientData = bridge.webViewClientData;
    globalThis.devvit = {
        context,
        dependencies: {
            client: queryClientVersion(document),
            webViewScripts: globalThis.webViewScriptsVersion,
        },
        entrypoints: clientData?.appConfig?.entrypoints ?? {},
        share: bridge?.shareParam ? { userData: bridge?.shareParam.userData } : undefined,
        appPermissionState,
        token: (bridge.signedRequestContext || bridge.webbitToken),
        webViewMode: bridge?.viewMode,
        startTime: bridge?.startTime ?? undefined,
        refreshToken: requestTokenRefresh,
    };
};
export function updateConsentStatus(consentStatus) {
    if (!globalThis.devvit.appPermissionState)
        return;
    globalThis.devvit.appPermissionState.consentStatus = consentStatus;
}
export function updateMode(mode) {
    devvit.webViewMode = mode;
}
export function updateToken(token) {
    globalThis.devvit.token = token;
}
/** @internal */
export function contextFromRequestContext(reqCtx, client, postData) {
    if (!reqCtx.app)
        throw Error('no RequestContext.app');
    if (!reqCtx.post)
        throw Error('no RequestContext.post');
    if (!reqCtx.subreddit)
        throw Error('no RequestContext.subreddit');
    return {
        appName: reqCtx.app.name,
        appSlug: reqCtx.app.slug,
        appVersion: reqCtx.app.version,
        client,
        commentId: undefined,
        loid: reqCtx.user?.devvitLoid ?? reqCtx.user?.devvit_loid,
        postAuthorId: T2(reqCtx.post.author),
        postData: postData?.developerData,
        postId: T3(reqCtx.post.id),
        snoovatar: reqCtx.user?.snoovatar,
        subredditId: T5(reqCtx.subreddit.id),
        subredditName: reqCtx.subreddit.name,
        userId: reqCtx.user?.id ? T2(reqCtx.user.id) : undefined,
        username: reqCtx.user?.name,
    };
}
/** @internal */
export function contextFromWebViewContext(webViewCtx, client, postData) {
    return {
        appName: webViewCtx.appName,
        appSlug: webViewCtx.appName,
        appVersion: webViewCtx.appVersion,
        client,
        commentId: undefined,
        loid: undefined,
        postData: postData?.developerData,
        postAuthorId: undefined,
        postId: T3(webViewCtx.postId),
        snoovatar: undefined,
        subredditId: T5(webViewCtx.subredditId),
        subredditName: webViewCtx.subredditName,
        userId: webViewCtx.userId ? T2(webViewCtx.userId) : undefined,
        username: undefined,
    };
}
/**
 * Extracts the BridgeContext from the hash.
 * @internal
 */
export function getBridgeContext(name, hash) {
    if (name)
        return JSON.parse(name);
    const json = hash?.slice(1); // Strip the leading '#' from the hash
    if (!json)
        throw Error('no bridge context');
    return JSON.parse(decodeURIComponent(json));
}
/** @internal */
export function getClient(bridge) {
    if ((bridge.client !== ClientName.ANDROID && bridge.client !== ClientName.IOS) ||
        !bridge.nativeVersion)
        return;
    return {
        name: bridge.client === ClientName.ANDROID ? 'ANDROID' : 'IOS',
        version: bridge.nativeVersion,
    };
}
