// @devvit/web-view-scripts/token.js
import { EffectType } from '@devvit/protos/json/devvit/ui/effects/v1alpha/effect.js';
import { emitEffectWithResponse } from '@devvit/shared-types/client/emit-effect.js';
import { jwtDecode } from 'jwt-decode';
import { updateToken } from './devvit-global.js';
export function decodeToken(token) {
    if (!token)
        return;
    try {
        return jwtDecode(token)?.devvit;
    }
    catch (err) {
        throw Error('token decode failure', { cause: err });
    }
}
export function initToken() {
    setInterval(refreshToken, 60000);
}
/** @internal */
export async function refreshToken() {
    // Check if the token is about to expire in the next 5 minutes
    const { exp: expiresAt } = jwtDecode(globalThis.devvit.token);
    if (expiresAt && expiresAt - Date.now() / 1000 > 300) {
        return;
    }
    await requestTokenRefresh();
}
export async function requestTokenRefresh() {
    const response = await emitEffectWithResponse({
        type: EffectType.EFFECT_UPDATE_REQUEST_CONTEXT,
        updateRequestContext: {},
    });
    if (!response?.updateRequestContext?.signedRequestContext) {
        console.error('Failed to refresh token');
        return;
    }
    updateToken(response.updateRequestContext.signedRequestContext);
}
