// @devvit/linkers/skinny-linker.js
import { assertNonNull } from '@devvit/shared-types/NonNull.js';
// to-do: It could be an option for linking in DefaultLinker that is passed. We
//        might also want a distinct type for skinny bundles. This Linker
//        literally doesn't link and resolve. There's no dependencies (uses).
export class SkinnyLinker {
    // Doesn't actually resolve and link dependencies because skinny bundles have
    // no dependencies.
    async link(bundle) {
        assertNonNull(bundle.dependencies?.actor, "Expected bundle's actor to be non-null.");
        // Generate everything needed to shape a client request / response.
        return {
            bundle: {
                // Intentionally omit code and uses. This is a skinny bundle. See
                // LinkedBundle.
                actor: bundle.dependencies.actor,
                buildInfo: bundle.buildInfo,
                code: '',
                hostname: bundle.dependencies.hostname,
                products: {},
                // This is what consumers of a skinny bundle use to shape their requests
                // and responses. It's a critical part of the API.
                provides: bundleProvidesSerializableServiceDefinition(bundle),
                // buildInfo intentionally omitted. This should be an implementation
                // detail.
                assets: bundle.assetIds,
                uses: [],
                webviewAssets: bundle.webviewAssetIds,
            },
            errors: [],
        };
    }
}
// Adapt a Bundle to the Bundle's provides as serializable service definitions.
function bundleProvidesSerializableServiceDefinition(bundle) {
    return (bundle.dependencies?.provides.flatMap((provide) => provide.definition ? [provide.definition] : []) ?? []);
}
