// @devvit/builders/Builder.js
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _DefaultBuilder_instances, _DefaultBuilder_pack, _DefaultBuilder_linker, _DefaultBuilder_link;
import { DefaultLinker } from '@devvit/linkers/linker.js';
// to-do: consolidate bundleApps implementations, move this file into it, and
//        delete the @devvit/builders package.
export class DefaultBuilder {
    constructor(pack, resolver) {
        _DefaultBuilder_instances.add(this);
        _DefaultBuilder_pack.set(this, void 0);
        _DefaultBuilder_linker.set(this, void 0);
        __classPrivateFieldSet(this, _DefaultBuilder_pack, pack, "f");
        __classPrivateFieldSet(this, _DefaultBuilder_linker, new DefaultLinker(resolver), "f");
    }
    async dispose() {
        await __classPrivateFieldGet(this, _DefaultBuilder_pack, "f").dispose();
    }
    async build(request) {
        const compileResponse = await __classPrivateFieldGet(this, _DefaultBuilder_pack, "f").compile(request);
        return await __classPrivateFieldGet(this, _DefaultBuilder_instances, "m", _DefaultBuilder_link).call(this, compileResponse);
    }
}
_DefaultBuilder_pack = new WeakMap(), _DefaultBuilder_linker = new WeakMap(), _DefaultBuilder_instances = new WeakSet(), _DefaultBuilder_link = async function _DefaultBuilder_link(rsp) {
    const linkResponses = await Promise.all(rsp.bundles?.map((bundle) => __classPrivateFieldGet(this, _DefaultBuilder_linker, "f").link(bundle)));
    const bundles = linkResponses.flatMap((response) => response.bundle ? [response.bundle] : []);
    const linkErrors = linkResponses?.flatMap((response) => response.errors.map((error) => ({ text: error.text })) ?? []);
    return {
        bundles,
        errors: [...rsp.errors, ...linkErrors],
        warnings: rsp.warnings,
    };
};
